### PR TITLE
Task 1: LinkedList

### PR DESCRIPTION
**Pull request by bryzhatov**

### CODE REVIEW

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                foundNode = foundNode.getNext();
+                builder.append(foundNode.getValue());
+                i++;
+            }
+        }
+        builder.append("]");
+        return builder.toString();
+    }
+
+    private class Node<E> {
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Why not static class?

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        Node<T> startNode = findNode(fromIndex);
+
+        for (int i = fromIndex; i <= toIndex; i++) {
+            subList.add(startNode.getValue());
+            startNode = startNode.getNext();
+        }
+        return subList;
+    }
+
+    public Node<T> getFirstElement() {
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Why it's a public method, which returns private class?

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || (index >= size && index != 0)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private void setLastModified() {
+        lastModified = System.currentTimeMillis();
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Why do you use the time? Any user can change the time and you will have a collision in iterator? Can you do it without system time?

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            }
+            find = find.getPrevious();
+            i--;
+        }
+        return -1;
+    }
+
+    @Override
+    public void clear() {
+        while (pastElement.hasPrevious()) {
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Why do you do this?

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+    public Node<T> getFirstElement() {
+        return firstElement;
+    }
+
+    public Node<T> getPastElement() {
+        return pastElement;
+    }
+
+    private Node<T> findNode(int foundIndexNode) {
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Would be nice to do it for O(n/2).

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private void setLastModified() {
+        lastModified = System.currentTimeMillis();
+    }
+
+    @Override
+    public String toString() {
```
Comment (jad7) [inline src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Why did you override this method? What you are using from the abstract class?

Comment (jad7) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: toArray() ?

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 1 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: It is resolved.

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 2 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: It is resolved.

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 4 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: Sorry for this. I resolved it.

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 5 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: It is resolved.

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 6 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: No, I just didn't know about exist toString() in AbstractCollection class :)

Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: I resolved this mistake.

File: src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java (hunk 3 above)
Comment (bryzhatov) [reply src/main/java/ua/griddynamics/trainee/task1/list/LinkedList.java]: It is resolved.

**Pull request merged**